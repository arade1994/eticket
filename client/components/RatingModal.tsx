import { useCallback, useState } from "react";
import Modal from "react-modal";
import { toast } from "react-toastify";

import { useRequest } from "../hooks/useRequest";
import RateInput from "./RateInput";

interface Props {
  open: boolean;
  currentUserId: string;
  selectedUserId: string;
  handleCloseClick: () => void;
  fetchRatings: () => Promise<any>;
}

const RatingModal = ({
  open,
  currentUserId,
  selectedUserId,
  handleCloseClick,
  fetchRatings,
}: Props) => {
  const [rate, setRate] = useState(5);
  const [comment, setComment] = useState("");
  const { sendRequest } = useRequest({
    url: !process.env.NEXT_PUBLIC_DEMO_MODE
      ? "/api/users/rate"
      : "http://localhost:3000/ratings",
    method: "post",
    body: { rate, comment, userId: currentUserId, ratedUserId: selectedUserId },
    onSuccess: () => {},
  });

  const handleSave = useCallback(async () => {
    await sendRequest();
    handleCloseClick();
    toast.success("User rated");
    fetchRatings();
  }, [sendRequest, handleCloseClick]);

  return (
    <Modal
      isOpen={open}
      appElement={document.createElement("div")}
      style={{
        content: {
          width: 400,
          height: 400,
          margin: "auto",
          display: "flex",
          flexDirection: "column",
          gap: 20,
        },
      }}
    >
      <div
        style={{
          display: "flex",
          flexDirection: "row",
          justifyContent: "space-between",
        }}
      >
        <div>
          <h3>Edit user rating</h3>
        </div>
        <div
          style={{ fontSize: 28, cursor: "pointer" }}
          onClick={handleCloseClick}
        >
          <p>&times;</p>
        </div>
      </div>
      <div
        style={{
          width: "100%",
          display: "flex",
          flexDirection: "column",
          justifyContent: "space-evenly",
          textAlign: "center",
        }}
      >
        <RateInput
          comment={comment}
          setComment={setComment}
          setRate={setRate}
        />
      </div>
      <div
        style={{
          display: "flex",
          flexDirection: "row",
          justifyContent: "space-around",
        }}
      >
        <button
          style={{
            borderRadius: 5,
            background: "rgba(7, 64, 82, 1)",
            color: "white",
            fontSize: 20,
          }}
          onClick={handleSave}
        >
          Save rating
        </button>
        <button
          style={{
            borderRadius: 5,
            fontSize: 20,
            background: "white",
          }}
          onClick={handleCloseClick}
        >
          Close
        </button>
      </div>
    </Modal>
  );
};

export default RatingModal;
