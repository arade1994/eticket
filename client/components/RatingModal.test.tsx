import { render } from "@testing-library/react";
import { describe, expect, test, vi } from "vitest";

import RatingModal from "./RatingModal";

const mockCurrentUserId = "currentUserId";
const mockSelectedUserId = "selectedUserId";

const mockOnClose = vi.fn();
const mockFetchRatings = vi.fn();

describe("<RatingModal />", () => {
  test("It should render rating modal as expected with all fields", () => {
    const { baseElement } = render(
      <RatingModal
        open={true}
        currentUserId={mockCurrentUserId}
        selectedUserId={mockSelectedUserId}
        handleCloseClick={mockOnClose}
        fetchRatings={mockFetchRatings}
      />
    );

    expect(baseElement).toMatchSnapshot();
  });
});
