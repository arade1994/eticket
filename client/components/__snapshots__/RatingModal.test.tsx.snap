// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<RatingModal /> > It should render rating modal as expected with all fields 1`] = `
<body
  class="ReactModal__Body--open"
>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div />
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div
    class="ReactModalPortal"
  >
    <div
      class="ReactModal__Overlay"
      style="position: fixed; top: 0px; left: 0px; right: 0px; bottom: 0px; background-color: rgba(255, 255, 255, 0.75);"
    >
      <div
        aria-modal="true"
        class="ReactModal__Content"
        role="dialog"
        style="position: absolute; top: 40px; left: 40px; right: 40px; bottom: 40px; border: 1px solid #ccc; background: rgb(255, 255, 255); overflow: auto; border-radius: 4px; outline: none; padding: 20px; width: 400px; height: 400px; margin: auto; display: flex; flex-direction: column; gap: 20px;"
        tabindex="-1"
      >
        <div
          style="display: flex; flex-direction: row; justify-content: space-between;"
        >
          <div>
            <h3>
              Edit user rating
            </h3>
          </div>
          <div
            style="font-size: 28px; cursor: pointer;"
          >
            <p>
              ×
            </p>
          </div>
        </div>
        <div
          style="width: 100%; display: flex; flex-direction: column; justify-content: space-evenly; text-align: center;"
        >
          <div
            class="rating"
          >
            <input
              id="star4"
              name="rating"
              type="radio"
              value="4"
            />
            <label
              class="star"
              for="star4"
            >
              ★
            </label>
            <input
              id="star3"
              name="rating"
              type="radio"
              value="3"
            />
            <label
              class="star"
              for="star3"
            >
              ★
            </label>
            <input
              id="star2"
              name="rating"
              type="radio"
              value="2"
            />
            <label
              class="star"
              for="star2"
            >
              ★
            </label>
            <input
              id="star1"
              name="rating"
              type="radio"
              value="1"
            />
            <label
              class="star"
              for="star1"
            >
              ★
            </label>
            <input
              id="star0"
              name="rating"
              type="radio"
              value="0"
            />
            <label
              class="star"
              for="star0"
            >
              ★
            </label>
          </div>
          <div
            class="comment"
          >
            <label
              for="comment"
            >
              Tell us more:
            </label>
            <br />
            <textarea
              id="comment"
              name="comment"
            />
          </div>
        </div>
        <div
          style="display: flex; flex-direction: row; justify-content: space-around;"
        >
          <button
            style="border-radius: 5px; background: rgb(7, 64, 82); color: white; font-size: 20px;"
          >
            Save rating
          </button>
          <button
            style="border-radius: 5px; font-size: 20px; background: white;"
          >
            Close
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;
